darts/models/regression_model.py

```diff
@@ -13,300 +13,289 @@
-        """
-        forecasts = self.historical_forecasts(feature_series,
-                                              target_series,
-                                              start,
-                                              forecast_horizon,
-                                              stride,
-                                              overlap_end,
-                                              last_points_only,
-                                              verbose)
-
-        if last_points_only:
-            return metric(target_series, forecasts)
-
-        errors = []
-        for forecast in forecasts:
-            errors.append(metric(target_series, forecast))
-
-        if reduction is None:
-            return errors
-
-        return reduction(errors)
-
-    def residuals(self) -> TimeSeries:
-        """ Computes the time series of residuals of this model on the training time series
-
-        The residuals are computed as
-
-        .. math:: z_t := y_t - \\hat{y}_t,
-
-        where :math:`y_t` is the actual target time series over the training set,
-        and :math:`\\hat{y}_t` is the time series of predicted targets, over the training set.
+        n : int
+            Forecast horizon - the number of time steps after the end of the series for which to produce predictions.
+        exog : Optional[TimeSeries], optional
+            The time series of exogenous variables which can be fed as input to the model. It must correspond to the
+            exogenous time series that has been used with the `fit()` method for training. It also must be of length `n`
+            and start one time step after the end of the training series.
+        **kwargs
+            Additional keyword arguments passed to the `predict` method of the model.
 
         Returns
         -------
         TimeSeries
-            The time series containing the residuals
+            A time series containing the `n` next points after then end of the training series.
         """
-
-        if (not self._fit_called):
-            raise_log(Exception('fit() must be called before predict()'), logger)
-
-        train_pred = self.predict(self.train_features)
-        return self.train_target - train_pred
+        super().predict(n, exog)
+
+        if self.max_lag != 0:
+            prediction_data = self.prediction_data.copy()
+            dummy_row = np.zeros(shape=(1, prediction_data.width))
+            prediction_data = prediction_data.append_values(dummy_row)
+
+        if exog is not None:
+            required_start = self.training_series.end_time()+self.training_series.freq()
+            raise_if_not(exog.start_time() == required_start,
+                "`exog` first date must be equal to self.training_series.end_time()+1*freq. " +
+                "Given: {}. Needed: {}.".format(exog.start_time(), required_start)
+            )
+        if isinstance(exog, TimeSeries):
+            exog = exog.pd_dataframe()
```

Comment:
maybe you could use `copy=False` here, as it doesn't seem you are touching `exog` in this code, and it might make it (slightly) faster to avoid copying.

Reply:
Done